File: day03/part2/main.dart
// -*- compile-command: "dart main.dart ../example1.txt ../input.txt"; -*-

import 'dart:io';
import 'package:tuple/tuple.dart';

main(List<String> args) {
  for (var arg in args) {
    process(arg);
  }

  print('Done.');
}

process(String filename) {
  print('Processing $filename ...');

	var rights = [1, 3, 5, 7, 1];
	var downs  = [1, 1, 1, 1, 2];

  var puz = Puzzle(filename);

  var result = 1;
	for (var i = 0; i < rights.length; i++) {
		var count = puz.countTrees(rights[i], downs[i]);
		result *= count;
	}

  print('Solution: $result');
}

class Puzzle {
  int width = 0;
  int height = 0;
  Set<Tuple2<int,int>> grid = Set<Tuple2<int,int>>();

  Puzzle(String filename) {
    var lines = File(filename).readAsLinesSync();
    for (var line in lines) {
      this.width = line.length;
      var chars = line.split('');
      for (var x = 0; x < chars.length; x++) {
        if (chars[x] == '#') {
          this.grid.add(Tuple2<int,int>(x, this.height));
        }
      }
      this.height++;
    }
  }

  int countTrees(int right, int down) {
    var posX = 0, posY = 0, count = 0;
    for (var y = 0; y < this.height; y++) {
      if (this.grid.contains(Tuple2<int,int>(posX % this.width, posY))) {
        count++;
      }
      posX += right;
      posY += down;
    }
    print('Found $count trees');
    return count;
  }
}

File: day03/part2/main.go
package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"strconv"
	"strings"
)

var (
	verbose = flag.Bool("v", false, "Verbose log messages")

	rights = []int{1, 3, 5, 7, 1}
	downs  = []int{1, 1, 1, 1, 2}
)

func main() {
	flag.Parse()

	for _, arg := range flag.Args() {
		process(arg)
	}

	log.Printf("Done.")
}

func process(filename string) {
	puz := readPuzzle(filename)

	result := 1
	for i := 0; i < len(rights); i++ {
		count := puz.countTrees(rights[i], downs[i])
		result *= count
	}

	log.Printf("Solution: %v", result)
}

func (p *Puzzle) countTrees(right, down int) int {
	var posX, posY, count int
	for y := 0; y < p.height; y++ {
		if p.lookup(posX, posY) {
			count++
		}
		posX += right
		posY += down
	}

	log.Printf("(%v,%v): Found %v trees", right, down, count)
	return count
}

type Puzzle struct {
	width  int
	height int
	grid   map[string]bool
}

func readPuzzle(filename string) *Puzzle {
	logf("Processing %v ...", filename)
	buf, err := ioutil.ReadFile(filename)
	check("ReadFile: %v", err)

	puz := &Puzzle{grid: map[string]bool{}}
	lines := strings.Split(string(buf), "\n")
	for y, line := range lines {
		if line == "" {
			continue
		}
		puz.width = len(line)
		puz.height++
		for x := 0; x < len(line); x++ {
			if line[x:x+1] == "#" {
				key := genKey(x, y)
				puz.grid[key] = true
			}
		}
	}

	return puz
}

func (p *Puzzle) lookup(x, y int) bool {
	u := x % p.width
	// v := y % p.height
	return p.grid[genKey(u, y)]
}

func genKey(x, y int) string {
	return fmt.Sprintf("%v,%v", y, x)
}

func parseKey(key string) (x, y int) {
	parts := strings.Split(key, ",")
	var err error
	x, err = strconv.Atoi(parts[1])
	check("key=%q parseKey(%q).x: %v", key, parts[1], err)
	y, err = strconv.Atoi(parts[0])
	check("key=%q parseKey(%q).y: %v", key, parts[0], err)
	return x, y
}

func check(fmtStr string, args ...interface{}) {
	if err := args[len(args)-1]; err != nil {
		log.Fatalf(fmtStr, args...)
	}
}

func logf(fmt string, args ...interface{}) {
	if *verbose {
		log.Printf(fmt, args...)
	}
}
